packages/dd-trace/src/telemetry/metrics.js

```diff
@@ -0,0 +1,219 @@
+}
+
+function mapToJsonArray (map) {
+  return Array.from(map.values()).map(v => v.toJSON())
+}
+
+class Metric {
+  constructor (namespace, metric, common, tags) {
+    this.namespace = namespace.toString()
+    this.metric = common ? metric : `nodejs.${metric}`
+    this.tags = tagArray(tags)
+    this.common = common
+
+    this.points = []
+  }
+
+  toString () {
+    const { namespace, metric } = this
+    return `${namespace}.${metric}`
+  }
+
+  reset () {
+    this.points = []
+  }
+
+  track () {
+    throw new Error('not implemented')
+  }
+
+  toJSON () {
+    const { metric, points, interval, type, tags, common } = this
+    return {
+      metric,
+      points,
+      interval,
+      type,
+      tags,
+      common
+    }
+  }
+}
+
+class CountMetric extends Metric {
+  get type () {
+    return 'count'
+  }
+
+  inc (value) {
+    return this.track(value)
+  }
+
+  dec (value = -1) {
+    return this.track(value)
+  }
+
+  track (value = 1) {
+    if (this.points.length) {
+      this.points[0][1] += value
+    } else {
+      this.points.push([now(), value])
+    }
+  }
+}
+
+class GaugeMetric extends Metric {
+  get type () {
+    return 'gauge'
+  }
+
+  mark (value) {
+    return this.track(value)
+  }
+
+  track (value = 1) {
+    this.points.push([now(), value])
+  }
+}
+
+class RateMetric extends Metric {
+  constructor (namespace, metric, common, tags, interval) {
+    super(namespace, metric, common, tags)
+
+    this.interval = interval
+    this.rate = 0
+  }
+
+  get type () {
+    return 'rate'
+  }
+
+  reset () {
+    super.reset()
+    this.rate = 0
+  }
+
+  track (value = 1) {
+    this.rate += value
+    const rate = this.interval ? (this.rate / this.interval) : 0.0
+    this.points = [[now(), rate]]
+  }
+}
+
+const metricsTypes = {
+  count: CountMetric,
+  gauge: GaugeMetric,
+  rate: RateMetric
+}
+
+class Namespace extends Map {
+  constructor (namespace) {
+    super()
+    this.namespace = namespace
+  }
+
+  reset () {
+    for (const metric of this.values()) {
+      metric.reset()
+    }
+  }
+
+  toString () {
+    return `dd.instrumentation_telemetry_data.${this.namespace}`
+  }
+
+  getMetric (type, name, tags, interval) {
+    const metricId = getId(type, this, name, tags)
+
+    let metric = this.get(metricId)
+    if (metric) return metric
+
+    const Factory = metricsTypes[type]
+    if (!Factory) {
+      throw new Error(`Unknown metric type ${type}`)
+    }
+
+    metric = new Factory(this, name, true, tags, interval)
+    this.set(metricId, metric)
+
+    return metric
+  }
+
+  count (name, tags) {
+    return this.getMetric('count', name, tags)
+  }
+
+  gauge (name, tags) {
+    return this.getMetric('gauge', name, tags)
+  }
+
+  rate (name, interval, tags) {
+    return this.getMetric('rate', name, tags, interval)
+  }
+
+  toJSON () {
+    const { namespace } = this
+    return {
+      namespace,
+      lib_language: 'nodejs',
+      lib_version: process.version,
```

Comment:
Why are you adding `lib_language` and `lib_version` properties to the payload? 
Shouldn't be added as tags? and shouldn't be `version` instead `lib_version`?

According to the doc:
- Common metrics will have the library’s `lib_language` and `version` as `tags.`
- Language specific metrics will have their respective `lib_version` as `tag`

Reply:
v1 expects them as part of the payload and this is designed for v1 as v2 is not ready yet and this needs to be out before that is ready.

Reply:
Ah, yep, fixed now. :)